src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt

```diff
@@ -25,42 +32,77 @@ fun <T : ILicenceCondition> conditionChanges(current: List<T>, other: List<T>) =
+    presentConditions.any { present -> current.code == present.code }
+  }
+  val changes = conditionChanges(conditionsToCheck, previousConditions, allReplacements)
 
-  conditions.filter { it.requiresInput }.forEach {
-    val placeholders = placeholderPattern.findAll(it.text).iterator()
-    val variables = variablePattern.findAll(it.tpl ?: "").toSet().iterator()
-    val keyValuePairs = HashMap<String, String>()
+  return changes.map { change -> change.copy(sequence = presentConditions.find { condition -> condition.code == change.code }?.sequence) }
+}
 
-    while (placeholders.hasNext() && variables.hasNext()) {
-      keyValuePairs[placeholders.next().value] = variables.next().value
-    }
+fun <T : Any> conditionChanges(
+  previousConditions: List<IAdditionalCondition<T>>,
+  currentConditions: List<IAdditionalCondition<T>>,
+  allReplacements: List<Replacements>
+): List<LicenceConditionChanges> = previousConditions.mapNotNull { previous ->
+  val current = currentConditions.find { it.code == previous.code }
+  if (current != null)
+    conditionUpdated(previous, current)
+  else
+    conditionRemoved(previous, allReplacements)
+}
 
-    results[it.code] = keyValuePairs
+private fun <T : Any> conditionUpdated(
+  previous: IAdditionalCondition<T>,
+  current: IAdditionalCondition<T>,
+): LicenceConditionChanges? {
+  val textHasChanged = previous.text != current.text
+  val removed = (previous.inputs ?: emptyList()) - (current.inputs ?: emptyList()).toSet()
+  val added = (current.inputs ?: emptyList()) - (previous.inputs ?: emptyList()).toSet()
+  val dataHasChanged = removed.isNotEmpty() || added.isNotEmpty()
+  return when {
+    textHasChanged -> TEXT_CHANGE.update(previous, current)
+    dataHasChanged -> NEW_OPTIONS.update(previous, current, removed, added)
+    else -> null
   }
-
-  return results
 }
 
-fun <T : ILicenceCondition> licencePolicyChanges(
-  conditions: List<AdditionalCondition>,
-  policyConditions: List<T>,
-  policyPlaceholders: HashMap<String, HashMap<String, String>>
-): List<LicenceConditionChanges> = conditions.mapNotNull { c ->
-  policyConditions.find { it.code == c.code }?.let { pc ->
-    val textHasChanged = pc.text != c.text
-    val dataChanges = c.data.filter { d -> policyPlaceholders[c.code]?.containsValue(d.field) == false }
-    if (textHasChanged || dataChanges.isNotEmpty()) LicenceConditionChanges(
-      c.code!!, c.sequence, c.text!!, pc.text, dataChanges
-    )
-    else null
+private fun conditionRemoved(
+  condition: ILicenceCondition,
+  allReplacements: List<Replacements>,
+): LicenceConditionChanges {
+  val replacements = allReplacements.find { it.code === condition.code }
+  return when (replacements?.changeType) {
+    DELETED -> DELETED.removal(condition, replacements.alternatives)
+    REPLACED -> REPLACED.removal(condition, replacements.alternatives)
+    else -> REMOVED_NO_REPLACEMENTS.removal(condition, emptyList())
```

Comment:
This isn't handled by the frontend and we don't have an example of this yet but it seemed quite an important thing to cover